onnx/defs/tensor/defs.cc

```diff
@@ -327,7 +452,31 @@ Example 2:
     .TypeConstraint(
         "Tind",
         {"tensor(int32)", "tensor(int64)"},
-        "Constrain indices to integer types");
+        "Constrain indices to integer types")
+    .ShapeInferenceFunction([](InferenceContext& ctx) {
+        if (!hasExactlyNInputs(ctx, 2, "Gather")) {
+          return;
+        }
+
+        propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+        if (!ctx.getInputType(0)->has_shape() ||
+            !ctx.getInputType(1)->has_shape()) {
+          return;
+        }
+
+        int r = ctx.getInputType(0)->shape()->dim_size();
+        int q = ctx.getInputType(1)->shape()->dim_size();
+
+        int out_rank = q + r - 1;
+
+        if (out_rank == 0) {
+          ctx.getOutputType(0)->mutable_shape();
+        }
+        for (int i = 0; i < out_rank; ++i) {
+          ctx.getOutputType(0)->mutable_shape()->add_dim();
```

Comment:
Is this supposed to be initialized to some value from the input dims?

Reply:
not at the moment - this is more "rank inference" than "shape inference". It can be enhanced in a future diff

Reply:
It would be safer to initialize the dim with the empty string param than leave it uninitialized (neither value nor param set).

Reply:
I guess representing this is still a little up in air - but see the comment I just wrote at https://github.com/onnx/onnx/pull/681#issuecomment-381221996

Reply:
Okay. As long as the meaning/representation is standardized and well-documented, it's fine.